### PR TITLE
Updates for upcoming sabre/xml 1.3 release

### PR DESCRIPTION
**Pull request by evert**
cc @staabm

### CODE REVIEW

File: source/xml/reading.md [hunk 1]
```diff
@@ -31,273 +44,318 @@ Let's take the following XML as our primary example.
 
 To convert this XML to a PHP array, we can just run this:
 
-    $reader = new Sabre\Xml\Reader();
-    $reader->xml($xml);
+    $service = new Sabre\Xml\Service();
 
-    print_r($reader->parse());
+    print_r($service->parse());
```
Comment (staabm) [inline source/xml/reading.md]: Xml arg

File: source/xml/reading.md [hunk 2]
```diff
@@ -522,29 +522,148 @@ parse themself from an XML document, and also write them in a new document:
...
+
+
+Using ValueObjects for this instead
+-----------------------------------
+
+While the former classmapping example is a good way to learn how to write custom
+deserializers, using the 'value objects' feature this could have been simplified
+even further.
+
+The exact same result could have been achieved by regestering the PHP classes as
```
Comment (staabm) [inline source/xml/reading.md]: Registering

File: source/xml/reading.md [hunk 3]
```diff
@@ -522,29 +522,148 @@ parse themself from an XML document, and also write them in a new document:
...
+`Sabre\Xml\XmlDeserializable`, you will end up with a function that receives
+an instance of `Sabre\Xml\Reader` such as this:
+
+
+    function myDeserializer(Sabre\Xml\Reader $reader) {
+
+    }
+
+The reader extends PHP's [XMLReader][1] object. You _must_ absolutely make
+sure that you read the _entire_ XML element, and not half way. The simples
```
Comment (staabm) [inline source/xml/reading.md]: simplest

File: source/xml/service.md [hunk 4]
```diff
@@ -22,99 +25,66 @@ This is the service class in birds eye view:
...
-         * This is the element map. It contains a list of XML elements (in clark
-         * notation) as keys and PHP class names as values.
-         *
-         * The PHP class names must implement Sabre\Xml\Element.
-         *
-         * Values may also be a callable. In that case the function will be called
-         * directly.
-         *
-         * @var array
+         * List of parsers for specifix XML objects.
```
Comment (staabm) [inline source/xml/service.md]: Specific

File: source/xml/service.md [hunk 5]
```diff
@@ -22,99 +25,66 @@ This is the service class in birds eye view:
...
+         */
+        function getWriter();
+
+        /**
+         * Maps an XML element to a PHP class.
+         */
+        function mapValueObject($elementName, $className);
+
+        /**
+         * Turn a mapped PHP object into an XML document
```
Comment (staabm) [inline source/xml/service.md]: Missing comment end

File: source/xml/valueobjects.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+To let sabre/xml automatically map between these two, simply use the Service
+class:
+
+    $service = new Sabre\Xml\Service();
+    $service->namespaceMap['http//www.w3.org/2005/Atom'] = 'atom';
+    
+    $service->mapValueObject('{http://www.w3.org/2005/Atom}feed', 'My\Atom\Feed'); 
+    $service->mapValueObject('{http://www.w3.org/2005/Atom}author', 'My\Atom\Author'); 
+    $service->mapValueObject('{http://www.w3.org/2005/Atom}entry', 'My\Atom\Entry');
```
Comment (staabm) [inline source/xml/valueobjects.md]: Maybe add a hint for users with a new enough php version to use `::class`?

Comment (evert) [inline source/xml/valueobjects.md]: Good idea =)

File: source/xml/valueobjects.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+the same XML namespace. If this is the case, we will see if the class
+`My\Atom\Entry` has a `public` property with the same name, and set it's value.
+
+If entry had sub-elements in a different XML namespace, they would be discarded.
+
+One special trick is that if you define your class with a property and give it
+a default value that's an array, `sabre/xml` will immediately assume that more
+than one element may appear. In the above example, both `$authors` and `$entry`
+was defaulted to an empty array. This signals `sabre/xml` that multiple
+`<author>` and `<entry>` elements may appear as children.
```
Comment (staabm) [inline source/xml/valueobjects.md]: And append those.

File: source/xml/valueobjects.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        public $length;
+
+    }
+
+And now, a custom deserializer, defined on the Service:
+
+    $service->elementMap['{http://www.w3.org/2005/Atom}link'] = function($reader) {
+
+        $link = new My\Atom\Link();
+        foreach($reader->parseAttributes() as $key=>$value) {
```
Comment (staabm) [inline source/xml/valueobjects.md]: Should we also provide a deserializer function which puts attributes onto a object properties?

Comment (evert) [inline source/xml/valueobjects.md]: Not a bad idea, but even if we have one... I think it's useful to have a 'fully written out' example in the documentation. I hope it helps people write custom deserializers.

I think in the real world we'll often see a mix between attributes and child-elements, and for those cases we can never really generalize things.

Comment (staabm) [inline source/xml/valueobjects.md]: Totally agree

Comment (staabm) [inline source/xml/valueobjects.md]: Just a idea: we could further break down the serializers and use iterators which takes a callable with all available attributes.

That way you could re-use it not matter if you have a object, array or some other structure

Comment (evert) [inline source/xml/valueobjects.md]: Sounds interesting =)

Comment (staabm) [inline source/xml/valueobjects.md]: I am thinking about something like

```php
$service->elementMap['{http://www.w3.org/2005/Atom}link'] = function($reader) {
  return Sabre\Xml\Deserializers\array($reader, function($array) {
    /*
    // represents the whole all elements and attributes
    $array === [
      ['tag1', 'attributes' => ['padding' => '5',...]]
      ['tag2', 'attributes' => ['padding' => '5',...]]
    ]
    */
    // e.g. I am only interessted in the padding attribute of tag1
    return $array[0]['tag1']['attributes']['padding']; // the deserializer func returns whatever this callable returns
  });
}
```

Comment (evert) [inline source/xml/valueobjects.md]: This sounds very similar to:

```
$service->elementMap['{http://www.w3.org/2005/Atom}link'] = function($reader) {
       $subTree = $reader->parseInnerTree();
  });
}
```

The array structure is a bit different, but this is exactly what you describe, no?

**Pull request merged**